File: reduks/src/main/java/com/beyondeye/reduks/Middleware.java
package com.beyondeye.reduks;

/**
 * keep this interface in java to leverage kotlin SAM conversion (lambda automatical converted to implementation of java interface)
 * see also https://github.com/reactjs/redux/blob/master/docs/Glossary.md#middleware
 */
public interface Middleware<S> {
    Object dispatch(Store<S> store, NextDispatcher next, Object action);
}

File: reduks/src/main/kotlin/com/beyondeye/reduks/Middleware.kt
package com.beyondeye.reduks

/**
 * single method interface, mainly used because kotlin does not support yet type alias for function types
 * see also https://github.com/reactjs/redux/blob/master/docs/Glossary.md#middleware
 */
interface Middleware<S> {
    fun dispatch(store: Store<S>, nextDispatcher:  (Any)->Any, action: Any): Any
}
